brave/src/test/java/brave/sampler/RateLimitingSamplerSoakTest.java

```diff
@@ -0,0 +1,66 @@
+package brave.sampler;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import org.junit.experimental.theories.DataPoints;
+import org.junit.experimental.theories.Theories;
+import org.junit.experimental.theories.Theory;
+import org.junit.runner.RunWith;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assumptions.assumeThat;
+
+@RunWith(Theories.class)
+public class RateLimitingSamplerSoakTest {
+
+  @DataPoints public static final int[] SAMPLE_RESERVOIRS = {11, 101, 1001, 1_000_001};
+
+  /** This test will take a little over a second per reservoir */
+  @Theory public void retainsPerSampleRate(int reservoir) throws Exception {
```

Comment:
I refactored this a little bit as I noticed our previous test didn't guarantee it covered the second. Still I think maybe there is a hidden problem we aren't able to see, yet.

advice on how to test this better welcome!

Reply:
Would it help to inject a clock? I notice the guava tests are easy to reason about since they don't use real time. They probably run way faster too

https://github.com/google/guava/blob/0cd4e9faa1360da4a343f84cb275d6eda0c5e732/android/guava-tests/test/com/google/common/util/concurrent/RateLimiterTest.java